util/HalideTraceDump.cpp

```diff
@@ -0,0 +1,403 @@
+#include "HalideBuffer.h"
+#include "halide_image_io.h"
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <stdint.h>
+#include <vector>
+#include <map>
+#include <assert.h>
+#include <string>
+#include <unistd.h>
+#include <fcntl.h>
+#include <string.h>
+
+/** \file
+ *
+ * A tool which can read a binary Halide trace file from stdin, and dump files containing the final pixel values
+ * recorded for each traced Func.
+ *
+ * Currently dumps into supported Halide image formats.
+ */
+
+using std::map;
+using std::vector;
+using std::string;
+using Halide::Runtime::Buffer;
+
+struct BufferOutputOpts {
+    enum OutputType {
+        PNG = 0,
+        JPG,
+        PGM
+    };
+
+    enum OutputType type;
+};
+
+int halide_copy_to_host(void *user_context, struct halide_buffer_t *buf) {
+    return 0;
+}
+
+struct Value {
+    uint8_t payload[8];
+    uint8_t defined[8];
+
+    Value() { defined[0] = 0; }
+    Value(Packet *p, int lane_idx=0) {
+        defined[0] = 1;
+        int bytes = p->type.bytes();
+        unsigned long addr = (unsigned long) p->value();
+        addr += (bytes*lane_idx);
+        memcpy((void*)payload, (void*)addr, bytes);
+    }
+
+    template<typename T>
+    T get_as(halide_type_t type) {
+        return value_as<T>(type, (void*)payload);
+    }
+};
+
+struct FuncInfo {
+    int min_coords[16];
+    int max_coords[16];
+    int extents[16];
+    int dimensions;
+    halide_type_t type;
+    Value * values;
+
+    FuncInfo() : values(NULL) {}
+    FuncInfo(Packet *p) : values(NULL) {
+        int real_dims = p->dimensions/p->type.lanes;
+        assert(real_dims <= 16);
+        for(int i=0; i<real_dims; i++){
+            min_coords[i] = INT32_MAX;
+            max_coords[i] = INT32_MIN;
+        }
+        dimensions = real_dims;
+        type = p->type;
+        type.lanes = 1;
+    }
+    ~FuncInfo() {
+        if(values != NULL) {
+            free(values);
+        }
+    }
+
+    void add_preprocess(Packet *p){
+        int real_dims = p->dimensions/p->type.lanes;
+        int lanes = p->type.lanes;
+
+        halide_type_t scalar_type = p->type;
+        scalar_type.lanes = 1;
+
+        assert(scalar_type == type);
+        assert(real_dims == dimensions);
+
+        for(int lane=0; lane < lanes; lane++) {
+            for(int i=0; i<real_dims; i++) {
+                if(p->coordinates()[lanes*i+lane] < min_coords[i]) { min_coords[i] = p->coordinates()[lanes*i+lane]; }
+                if(p->coordinates()[lanes*i+lane] > max_coords[i]) { max_coords[i] = p->coordinates()[lanes*i+lane]; }
+            }
+        }
+    }
+
+    void allocate() {
+        int size = 1;
+        for(int i=0; i<dimensions; i++) {
+            extents[i] = max_coords[i] - min_coords[i] + 1;
+            size *= extents[i];
+        }
+        values = (Value *)malloc(size * sizeof(Value));
+        for(int i=0; i<size; i++) {
+            values[i] = Value();
+        }
+        assert(values != NULL);
```

Comment:
I opted for a two-pass implementation because trace files can be immense. The first pass is just to determine what the ranges are of all buffer extents, so that allocation can be at once before the second pass, as opposed to slow dynamic allocation for each new element encountered.

I thought (maybe wrongly) that using nested vectors for each dimension would be slow to access. 

It could still use a single-dimensional vector. If you think that is better, please confirm and I'll update it.